File: Arbimaps/vistas/tramites/base_catastral/generar_indice_predial.py
import csv
import json
import time
from pathlib import Path


def col_name(index):
    index += 1
    name = ""
    while index:
        index, remainder = divmod(index - 1, 26)
        name = chr(65 + remainder) + name
    return name


def read_grouped(path):
    grouped = {}
    with path.open("r", encoding="utf-8-sig", newline="") as handle:
        reader = csv.reader(handle, delimiter=";")
        raw_headers = next(reader)
        headers = {
            col_name(index): (header.strip() or f"Columna {col_name(index)}")
            for index, header in enumerate(raw_headers)
        }

        for line_number, row in enumerate(reader, start=2):
            if len(row) < 3:
                continue

            numero_predial = row[2].strip()
            if not numero_predial:
                continue

            values = {
                col_name(index): (row[index].strip() if index < len(row) else "")
                for index in range(len(raw_headers))
            }

            item = grouped.setdefault(numero_predial, {"headers": headers, "rows": []})
            item["rows"].append({"row_number": line_number, "values": values})

    return grouped


def main():
    root = Path(__file__).resolve().parents[4]
    csv1 = root / "dat_neiva" / "CATASTRAL_REGISTRO1_20250505.csv"
    csv2 = root / "dat_neiva" / "CATASTRAL_REGISTRO2_20250505.csv"
    output_dir = root / "dat_neiva" / "indice_predial"
    output_dir.mkdir(parents=True, exist_ok=True)

    start = time.time()
    print("Leyendo Registro 1...")
    registro1 = read_grouped(csv1)
    print(f"Registro 1: {len(registro1)} predios")

    print("Leyendo Registro 2...")
    registro2 = read_grouped(csv2)
    print(f"Registro 2: {len(registro2)} predios")

    predios = sorted(set(registro1) | set(registro2))
    metadata = {
        "version": "indice-predial-v1",
        "generated_at": time.strftime("%Y-%m-%d %H:%M:%S"),
        "registro1_csv": csv1.name,
        "registro2_csv": csv2.name,
        "total_predios": len(predios),
    }

    (output_dir / "metadata.json").write_text(
        json.dumps(metadata, ensure_ascii=False, indent=2),
        encoding="utf-8",
    )

    for index, numero_predial in enumerate(predios, start=1):
        bucket = output_dir / (numero_predial[:6] or "sin_id")
        bucket.mkdir(parents=True, exist_ok=True)

        payload = {
            "numero_predial": numero_predial,
            "registro1": registro1.get(numero_predial, {"headers": {}, "rows": []}),
            "registro2": registro2.get(numero_predial, {"headers": {}, "rows": []}),
        }

        (bucket / f"{numero_predial}.json").write_text(
            json.dumps(payload, ensure_ascii=False, separators=(",", ":")),
            encoding="utf-8",
        )

        if index % 10000 == 0:
            print(f"Escritos {index}/{len(predios)}")

    elapsed = time.time() - start
    print(f"Indice generado en {output_dir} con {len(predios)} predios en {elapsed:.1f}s")


if __name__ == "__main__":
    main()

File: Arbimaps/vistas/tramites/base_catastral/generar_indice_predial.php
<?php
// Ejecutar una vez cuando cambien los CSV oficiales.
// Genera un indice por numero predial para evitar recorrer CSV completos por cada radicado.

require_once __DIR__ . '/base_catastral_lib.php';

use PhpOffice\PhpSpreadsheet\Cell\Coordinate;

set_time_limit(0);
ini_set('memory_limit', '1024M');

function bci_read_csv_grouped(string $csvPath, string $registro): array
{
    $handle = fopen($csvPath, 'r');
    if (!$handle) {
        throw new RuntimeException("No se pudo abrir $csvPath");
    }

    $headerRow = fgetcsv($handle, 0, ';');
    if (!is_array($headerRow)) {
        fclose($handle);
        throw new RuntimeException("No se pudo leer encabezado de $csvPath");
    }

    $headers = [];
    foreach ($headerRow as $index => $header) {
        $col = Coordinate::stringFromColumnIndex($index + 1);
        $headers[$col] = bc_normalize_value($header) ?: 'Columna ' . $col;
    }

    $grouped = [];
    $lineNumber = 1;
    while (($data = fgetcsv($handle, 0, ';')) !== false) {
        $lineNumber++;
        $numeroPredial = bc_normalize_value($data[2] ?? '');
        if ($numeroPredial === '') {
            continue;
        }

        $values = [];
        foreach (array_keys($headers) as $index => $col) {
            $values[$col] = bc_normalize_value($data[$index] ?? '');
        }

        if (!isset($grouped[$numeroPredial])) {
            $grouped[$numeroPredial] = [
                'headers' => $headers,
                'rows' => [],
            ];
        }

        $grouped[$numeroPredial]['rows'][] = [
            'row_number' => $lineNumber,
            'values' => $values,
        ];
    }

    fclose($handle);
    return $grouped;
}

$root = bc_root_path();
$csv1 = $root . '/' . BC_REGISTRO1_CSV;
$csv2 = $root . '/' . BC_REGISTRO2_CSV;
$outDir = $root . '/dat_neiva/indice_predial';

if (!is_dir($outDir)) {
    mkdir($outDir, 0777, true);
}

$registro1 = bci_read_csv_grouped($csv1, 'registro1');
$registro2 = bci_read_csv_grouped($csv2, 'registro2');
$predios = array_unique(array_merge(array_keys($registro1), array_keys($registro2)));
sort($predios);

$meta = [
    'version' => 'indice-predial-v1',
    'generated_at' => date('Y-m-d H:i:s'),
    'registro1_csv' => basename($csv1),
    'registro2_csv' => basename($csv2),
    'total_predios' => count($predios),
];

file_put_contents($outDir . '/metadata.json', json_encode($meta, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));

$count = 0;
foreach ($predios as $numeroPredial) {
    $prefix = substr($numeroPredial, 0, 6) ?: 'sin_id';
    $bucketDir = $outDir . '/' . $prefix;
    if (!is_dir($bucketDir)) {
        mkdir($bucketDir, 0777, true);
    }

    $payload = [
        'numero_predial' => $numeroPredial,
        'registro1' => $registro1[$numeroPredial] ?? ['headers' => [], 'rows' => []],
        'registro2' => $registro2[$numeroPredial] ?? ['headers' => [], 'rows' => []],
    ];

    file_put_contents($bucketDir . '/' . $numeroPredial . '.json', json_encode($payload, JSON_UNESCAPED_UNICODE));
    $count++;
}

echo "Indice generado en $outDir con $count predios.\n";
